lib/configgin.rb

```diff
@@ -58,73 +57,86 @@ def generate_jobs(job_configs, templates)
+  def export_job_properties(jobs)
+    # co-located containers don't get to export properties
+    return unless instance_group == ENV["KUBERNETES_CONTAINER_NAME"]
+    # jobs don't export properties
+    return unless self_pod['metadata']['ownerReferences'][0]['kind'] == "StatefulSet"
 
+    sts = kube_client_stateful_set.get_stateful_set(instance_group, kube_namespace)
+
+    # Make sure the secret attached to the stateful set exists.
+    # XXX This should probably be done by fissile via the helm chart.
     secret = Kubeclient::Resource.new
-    secret.metadata = {}
-    # Prefixing with pod.metadata.name is purely for human convenience/debugging.
-    secret.metadata.name = "#{pod.metadata.name}-#{pod.metadata.uid}"
-    secret.metadata.namespace = kube_namespace
-
-    # Make sure the secret gets removed when the pod is deleted.
-    secret.metadata.ownerReferences = [
-      {
-        apiVersion: pod.apiVersion,
-        blockOwnerDeletion: false,
-        controller: false,
-        kind: pod.kind,
-        name: pod.metadata.name,
-        uid: pod.metadata.uid,
-      }
-    ]
+    secret.metadata = {
+      name: sts.metadata.name,
+      namespace: kube_namespace,
+      ownerReferences:  [
+        {
+          apiVersion: sts.apiVersion,
+          blockOwnerDeletion: false,
+          controller: false,
+          kind: sts.kind,
+          name: sts.metadata.name,
+          uid: sts.metadata.uid,
+        }
+      ]
+    }
+    begin
+      kube_client.create_secret(secret)
+    rescue
+    end
+    secret = kube_client.get_secret(instance_group, kube_namespace)
+    secret.data ||= {}
+
+    version_tag = ENV["CONFIGGIN_VERSION_TAG"]
+    new_tag = !secret.data.has_key?(version_tag)
+    secret.data = {version_tag => ""} if new_tag # make sure old properties are deleted during upgrade
 
-    secret.data = {}
     digests = {}
     jobs.each do |name, job|
       digests[name] = property_digest(job.exported_properties)
       secret.data["skiff-exported-properties-#{name}"] = Base64.encode64(job.exported_properties.to_json)
-      secret.data["skiff-exported-digest-#{name}"] = Base64.encode64(digests[name])
-    end
 
-    # Only the main container gets to export properties; colocated sidecars don't.
-    if instance_group == ENV["KUBERNETES_CONTAINER_NAME"]
-      begin
-        kube_client.delete_secret(secret.metadata.name, kube_namespace)
-      rescue
+      encoded_digest = Base64.encode64(digests[name])
+
+      # Record initial digest values whenever the tag changes, in which case the pod startup
+      # order is already controlled by the "CONFIGGIN_IMPORT_#{role}" references to the new
+      # tags in the corresponding secrets. There is no annotation when importing this set of
+      # initial values because the helm chart doesn't include any annotations, and we don't
+      # want to trigger a pod restart by adding them.
+      if new_tag
+        secret.data["skiff-initial-digest-#{name}"] = encoded_digest
```

Comment:
```suggestion
      secret.data["skiff-initial-digest-#{name}"] = encoded_digest if new_tag
```

Reply:
I normally like using statement modifiers like this, but in this case I prefer to write the conditions first, to make it visually more obvious.